src/typing/flow_js.ml

```diff
@@ -4236,6 +4239,18 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
       let new_obj = ObjT (reason, {objtype with flags}) in
       rec_flow_t cx trace (new_obj, t)
 
+    (*********************************************************)
+    (* objects can be marked as read-only using $ReadOnly<T> *)
+    (*********************************************************)
+
+    | (ObjT (reason_o, objtype), UseT(_, ReadOnlyT (reason_op, t))) ->
+      let desc = RReadOnly (desc_of_reason reason_o) in
+      let reason = replace_reason_const desc reason_op in
+
+      let flags = {frozen = true; sealed = Sealed; exact = true;} in
+      let new_obj = ObjT (reason, {objtype with flags}) in
```

Comment:
What you are saying sounds sensible, but I'm not sure I know how to do that 😞 

At this stage I would really appreciate some help in moving this forward from someone with a bit more familiarity with the codebase.